graphql_compiler/compiler/ir_lowering_match/utils.py

```diff
@@ -60,6 +85,117 @@ def to_gremlin(self):
+    inner_location_existence = BinaryComposition(u'!=', inner_local_field, NullLiteral)
+
+    edge_field_location = Location(root_location_path, field=edge_field)
+    select_edge_context_field = SelectEdgeContextField(edge_field_location)
+    edge_field_non_existence = filter_edge_field_non_existence(select_edge_context_field)
+
+    return BinaryComposition(u'||', edge_field_non_existence, inner_location_existence)
+
+
+class SelectWhereFilter(Filter):
+    """A Filter object that is used for the WHERE clause in a SELECT statement."""
+
+    def __init__(self, simple_optional_root_info):
+        """Construct a Filter block that is True if and only if each simple optional filter is True.
+
+        Construct filters for each simple optional, that are True if and only if `edge_field` does
+        not exist in the `simple_optional_root_location` OR the `inner_location` is not defined.
+        Return a SelectWhereFilter that evaluates to True if and only if *all* of the
+        aforementioned filters evaluate to True (conjunction).
+
+        Args:
+            simple_optional_root_info: dict mapping from simple_optional_root_location -> dict
+                                       containing keys
+                                       - 'inner_location_name': Location object correspoding to the
+                                                                unique MarkLocation present within a
+                                                                simple @optional (one that does not
+                                                                expands vertex fields) scope
+                                       - 'edge_field': string representing the optional edge being
+                                                       traversed
+                                       where simple_optional_root_to_inner_location is the location
+                                       preceding the @optional scope
+        Returns:
+            a new SelectWhereFilter object
+        """
+        inner_location_name_to_where_filter = {}
+        for root_location, root_info_dict in six.iteritems(simple_optional_root_info):
+            inner_location_name = root_info_dict['inner_location_name']
+            edge_field = root_info_dict['edge_field']
+
+            optional_edge_where_filter = _filter_orientdb_simple_optional_edge(
+                root_location.query_path, inner_location_name, edge_field)
+            inner_location_name_to_where_filter[inner_location_name] = optional_edge_where_filter
+
+        # Sort expressions by inner_location_name to obtain deterministic order
+        where_filter_expressions = [
+            inner_location_name_to_where_filter[key]
+            for key in sorted(inner_location_name_to_where_filter.keys())
+        ]
+
+        predicate = expression_list_to_conjunction(where_filter_expressions)
+        super(SelectWhereFilter, self).__init__(predicate)
```

Comment:
Let's make this take a predicate expression instead, and move the logic elsewhere. I'm not a fan of having tons of complex logic in a constructor like this.